File: packages/oxlint-plugin-react-doctor/src/plugin/utils/is-react-router-session-method.ts
import type { SymbolDescriptor } from "../semantic/scope-analysis.js";
import { getImportedNameFromReactRouter } from "./get-imported-name-from-react-router.js";
import { getStaticPropertyKeyName } from "./get-static-property-key-name.js";
import { isNodeOfType } from "./is-node-of-type.js";
import type { RuleContext } from "./rule-context.js";

const SESSION_STORAGE_FACTORY_EXPORT_NAMES = new Set([
  "createCookieSessionStorage",
  "createMemorySessionStorage",
  "createSessionStorage",
]);

export const isReactRouterSessionMethod = (
  context: RuleContext,
  symbol: SymbolDescriptor | null,
  expectedMethodName: string,
): boolean => {
  if (symbol === null) return false;
  const property = symbol.bindingIdentifier.parent;
  if (!isNodeOfType(property, "Property")) return false;
  if (getStaticPropertyKeyName(property, { allowComputedString: true }) !== expectedMethodName) {
    return false;
  }
  if (!isNodeOfType(symbol.initializer, "CallExpression")) return false;
  if (!isNodeOfType(symbol.initializer.callee, "Identifier")) return false;
  const factoryName = getImportedNameFromReactRouter(
    context,
    symbol.initializer.callee,
    symbol.initializer.callee.name,
  );
  return factoryName !== null && SESSION_STORAGE_FACTORY_EXPORT_NAMES.has(factoryName);
};

File: native/oxlint/rules/is-react-router-session-method.rs
const REACT_ROUTER_SESSION_RUNTIME_PACKAGE_NAMES: [&str; 5] = [
    "@react-router/cloudflare",
    "@react-router/node",
    "react-router/dom",
    "react-router-dom",
    "react-router",
];
const REACT_ROUTER_SESSION_STORAGE_FACTORY_NAMES: [&str; 5] = [
    "createCookieSessionStorage",
    "createFileSessionStorage",
    "createMemorySessionStorage",
    "createSessionStorage",
    "createWorkersKVSessionStorage",
];

fn is_react_router_session_method(
    identifier: &oxc_ast::ast::IdentifierReference<'_>,
    expected_method_name: &str,
    ctx: &crate::context::LintContext<'_>,
) -> bool {
    let Some(symbol_id) = ctx
        .scoping()
        .get_reference(identifier.reference_id())
        .symbol_id()
    else {
        return false;
    };
    let declaration = ctx.symbol_declaration(symbol_id);
    let oxc_ast::AstKind::VariableDeclarator(declarator) = declaration.kind() else {
        return false;
    };
    let oxc_ast::ast::BindingPattern::ObjectPattern(pattern) = &declarator.id else {
        return false;
    };
    if !pattern.properties.iter().any(|property| {
        property.key.static_name().as_deref() == Some(expected_method_name)
            && binding_pattern_has_symbol(&property.value, symbol_id)
    }) {
        return false;
    }
    let Some(oxc_ast::ast::Expression::CallExpression(factory_call)) = &declarator.init else {
        return false;
    };
    let oxc_ast::ast::Expression::Identifier(factory_callee) =
        factory_call.callee.get_inner_expression()
    else {
        return false;
    };
    direct_named_import_matches(
        factory_callee,
        &REACT_ROUTER_SESSION_STORAGE_FACTORY_NAMES,
        &REACT_ROUTER_SESSION_RUNTIME_PACKAGE_NAMES,
        ctx,
    )
}

fn is_react_router_session_method_call(
    call_expression: &oxc_ast::ast::CallExpression<'_>,
    session_symbol_id: oxc_semantic::SymbolId,
    expected_method_name: &str,
    ctx: &crate::context::LintContext<'_>,
) -> bool {
    let oxc_ast::ast::Expression::Identifier(callee) = &call_expression.callee else {
        return false;
    };
    let Some(oxc_ast::ast::Expression::Identifier(session_argument)) = call_expression
        .arguments
        .first()
        .and_then(oxc_ast::ast::Argument::as_expression)
    else {
        return false;
    };
    ctx.scoping()
        .get_reference(session_argument.reference_id())
        .symbol_id()
        == Some(session_symbol_id)
        && is_react_router_session_method(callee, expected_method_name, ctx)
}
